File: Library/src/main/java/org/endeavourhealth/library/model/iml/UIProperty.java
package org.endeavourhealth.library.model.iml;

import java.util.ArrayList;
import java.util.List;
import org.endeavourhealth.library.model.tripletree.TTIriRef;

public class UIProperty {

  private String iri;
  private String name;
  private String propertyType;
  private String valueType;
  private String valueTypeName;
  private int maxCount;
  private int number;
  private String intervalUnitIri;
  private List<TTIriRef> intervalUnitOptions;
  private String unitIri;
  private List<TTIriRef> unitOptions;
  private String operatorIri;
  private List<String> operatorOptions;
  private List<TTIriRef> qualifierOptions;
  private Integer setMemberCount;

  public UIProperty() {}

  public String getIri() {
    return iri;
  }

  public UIProperty setIri(String iri) {
    this.iri = iri;
    return this;
  }

  public String getName() {
    return name;
  }

  public UIProperty setName(String name) {
    this.name = name;
    return this;
  }

  public String getPropertyType() {
    return propertyType;
  }

  public UIProperty setPropertyType(String propertyType) {
    this.propertyType = propertyType;
    return this;
  }

  public String getValueType() {
    return valueType;
  }

  public UIProperty setValueType(String valueType) {
    this.valueType = valueType;
    return this;
  }

  public int getMaxCount() {
    return maxCount;
  }

  public UIProperty setMaxCount(int maxCount) {
    this.maxCount = maxCount;
    return this;
  }

  public int getNumber() {
    return number;
  }

  public UIProperty setNumber(int number) {
    this.number = number;
    return this;
  }

  public String getIntervalUnitIri() {
    return intervalUnitIri;
  }

  public UIProperty setIntervalUnitIri(String intervalUnitIri) {
    this.intervalUnitIri = intervalUnitIri;
    return this;
  }

  public List<TTIriRef> getIntervalUnitOptions() {
    return intervalUnitOptions;
  }

  public UIProperty setIntervalUnitOptions(List<TTIriRef> intervalUnitOptions) {
    this.intervalUnitOptions = intervalUnitOptions;
    return this;
  }

  public String getUnitIri() {
    return unitIri;
  }

  public UIProperty setUnitIri(String unitIri) {
    this.unitIri = unitIri;
    return this;
  }

  public List<TTIriRef> getUnitOptions() {
    return unitOptions;
  }

  public UIProperty setUnitOptions(List<TTIriRef> unitOptions) {
    this.unitOptions = unitOptions;
    return this;
  }

  public String getOperatorIri() {
    return operatorIri;
  }

  public UIProperty setOperatorIri(String operatorIri) {
    this.operatorIri = operatorIri;
    return this;
  }

  public List<String> getOperatorOptions() {
    return operatorOptions;
  }

  public UIProperty setOperatorOptions(List<String> operatorOptions) {
    this.operatorOptions = operatorOptions;
    return this;
  }

  public List<TTIriRef> getQualifierOptions() {
    return qualifierOptions;
  }

  public UIProperty setQualifierOptions(List<TTIriRef> qualifierOptions) {
    this.qualifierOptions = qualifierOptions;
    return this;
  }

  public Integer getSetMemberCount() {
    return setMemberCount;
  }

  public UIProperty setSetMemberCount(Integer setMemberCount) {
    this.setMemberCount = setMemberCount;
    return this;
  }

  public String getValueTypeName() {
    return this.valueTypeName;
  }

  public UIProperty setValueTypeName(String valueTypeName) {
    this.valueTypeName = valueTypeName;
    return this;
  }

  public UIProperty addQualifierOption(String iri, String name) {
    if (this.qualifierOptions == null) {
      this.qualifierOptions = new ArrayList<>();
    }
    this.qualifierOptions.add(new TTIriRef(iri, name));
    return this;
  }
}

File: src/interfaces/UIProperty.ts
// export interface UIProperty {
//   iri: string;
//   name: string;
//   propertyType: "class" | "datatype" | "node";
//   valueType: string;
//   maxCount: number;
//   minCount: number;
//   valueLabel: string;
//   intervalUnitIri: string;
//   intervalUnitOptions: TTIriRef[];
//   unitIri: string;
//   unitOptions: TTIriRef[];
//   operatorIri: string;
//   operatorOptions: string[];
//   qualifierOptions: TTIriRef[];
//   setMemberCount: number;
// }
import z from "zod";

import { TTIriRefSchema } from "./TTIriRef";

export const UIPropertySchema = z.strictObject({
  iri: z.string(),
  name: z.string(),
  propertyType: z.enum(["class", "datatype", "node"]),
  valueType: z.string(),
  maxCount: z.number(),
  minCount: z.number(),
  valueLabel: z.string(),
  intervalUnitIri: z.string(),
  intervalUnitOptions: z.array(TTIriRefSchema).prefault([]),
  unitIri: z.string(),
  unitOptions: z.array(TTIriRefSchema).prefault([]),
  operatorIri: z.string(),
  operatorOptions: z.array(z.string()).prefault([]),
  qualifierOptions: z.array(TTIriRefSchema).prefault([]),
  setMemberCount: z.number()
});

export type UIProperty = z.output<typeof UIPropertySchema>;
